### PR TITLE
Tutorials for event and command handlers

### PR DESCRIPTION
**Pull request by Lawouach**
Each handler is part of the .atomist project and has a test.

They ran fine with the CLI 0.32.0.

### CODE REVIEW

File: .atomist/handlers/event/NewCommitPushedToDMAndGeneral.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    handle(event: Match<Commit, Commit>): Plan {
+        const commit: Commit = event.root();
+        const repo: Repo = commit.repo;
+        const url: string = `https://github.com/${repo.owner}/${repo.name}/commit/${commit.sha}`;
+
+        const plan = new Plan();
+
+        const message = new DirectedMessage(
+            `${commit.author.name} has pushed ${url}`,
+            new ChannelAddress("general"));
```
Comment (kipz) [inline .atomist/handlers/event/NewCommitPushedToDMAndGeneral.ts]: Should be `#general`

File: .atomist/handlers/event/NewCommitPushedToDMAndGeneral.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        plan.add(message);
+
+        const dm = new DirectedMessage(
+            `${commit.author.name} has pushed ${url}`,
+            new UserAddress("Bob"));
```
Comment (kipz) [inline .atomist/handlers/event/NewCommitPushedToDMAndGeneral.ts]: Should be `@Bob`

File: .atomist/tests/handlers/command/ListActiveIssuesSteps.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+When("the ListActiveIssues is invoked", (world: CommandHandlerScenarioWorld) => {
+    const handler = world.commandHandler("ListActiveIssues");
+    world.invokeHandler(handler, {
+        repo: "rug",
+        owner: "atomist"
+    });
+});
+
+Then("you get a plan which will retrieve the list of open and active issues", (world: CommandHandlerScenarioWorld) => {
+    const instruction = world.plan().instructions[0] as Respondable<Execute>;
```
Comment (kipz) [inline .atomist/tests/handlers/command/ListActiveIssuesSteps.ts]: So we are still exposing and using the JVM plan for testing? I _think_ we decided _not_ to do this!

File: .atomist/tests/handlers/event/NewCommitPushedToDMSteps.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    const commit: Commit = new Commit();
+    commit.withSha("1234");
+
+    const author: GitHubId = new GitHubId();
+    author.withName("Alice");
+    commit.withAuthor(author);
+
+    const repo: Repo = new Repo();
+    repo.withOwner("atomist");
+    repo.withName("rug");
```
Comment (kipz) [inline .atomist/tests/handlers/event/NewCommitPushedToDMSteps.ts]: not "Bob"?

File: .atomist/handlers/event/NewCommitPushedToDMAndGeneral.ts (hunk 1 above)
Comment (ddgenome) [inline .atomist/handlers/event/NewCommitPushedToDMAndGeneral.ts]: According to @slimslenderslacks, either works. Similar for the leading `@` on users.

File: docs/tutorials/custom-commit-notifications.md [hunk 5]
```diff
@@ -0,0 +1,123 @@
+Things happen all the time in your development workflow. Commits are pushed,
+builds are started, they fail or succeed. Team members, or even users, open and
+comment issues everyday.
```
Comment (kipz) [inline docs/tutorials/custom-commit-notifications.md]: `common on issues`

File: docs/tutorials/custom-commit-notifications.md [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    These tutorials aim at showing you the code that you could write to
+    automate event reactions. They assume you have a [Rug project][ugpj]. You
+    can easily add [Rug event handler][rugev] to that project via the public 
+    [AddTypeScriptEventHandler Rug editor][rugeditor] provided by Atomist.
+
+[rugev]: /user-guide/rug/handlers.md
+[ugpj]: /user-guide/rug/projects.md
+[rugeditor]: https://github.com/atomist/rug-editors#addtypescripteventhandler
+
+## Getting Notified from an User Commit
```
Comment (kipz) [inline docs/tutorials/custom-commit-notifications.md]: `a User` - it's the sound of the first letter, not whether or not it's a vowel.

File: docs/tutorials/custom-commit-notifications.md [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+The first event you will ask Atomist to react to is commits pushed by a user. To
```
Comment (kipz) [inline docs/tutorials/custom-commit-notifications.md]: plural/singular - `first event` vs `commits`

File: docs/tutorials/custom-commit-notifications.md [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+Below is such an event handler:
+
+```typescript linenums="1"
+{!../../.atomist/handlers/event/NewCommitPushedToGeneral.ts!}
+```
+
+An event handler tells Atomist what it is interested in via its
+[path expression][ugpxe] on line 10. Notice in particular how the handler
+filters only commit made by a given author on line 11 (here, Alice's commits).
```
Comment (kipz) [inline docs/tutorials/custom-commit-notifications.md]: commits

File: docs/tutorials/custom-commit-notifications.md [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+An event handler tells Atomist what it is interested in via its
+[path expression][ugpxe] on line 10. Notice in particular how the handler
+filters only commit made by a given author on line 11 (here, Alice's commits). 
+Finally, line 12 asks Atomist to also feed the handler with the commit's
+repository details.
+
+[ugpxe]: /user-guide/rug/path-expressions.md
+
+Once it gets called, the event handler builds a message to send to the 
+`general` channel of your team.
```
Comment (kipz) [inline docs/tutorials/custom-commit-notifications.md]: `#general`

File: docs/tutorials/custom-commit-notifications.md [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+the general channel of your team when Alice pushes her commits.
+
+Let's see now you could ask Atomist to send a direct message instead.
+
+```typescript linenums="1" hl_lines="22"
+{!../../.atomist/handlers/event/NewCommitPushedToDM.ts!}
+```
+
+As you can see, the only change is on line 22 where you provide the name of 
+the user to send the message to, here `Bob`.
```
Comment (kipz) [inline docs/tutorials/custom-commit-notifications.md]: `@Bob`

File: docs/tutorials/custom-commit-notifications.md [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+
+But wait, you can have it both ways and send a direct message and a 
+channel message:
+
+```typescript linenums="1" hl_lines="21 26 31 33"
+{!../../.atomist/handlers/event/NewCommitPushedToDMAndGeneral.ts!}
+```
+
+Event handlers can trigger multiple messages from a single event as shown
```
Comment (kipz) [inline docs/tutorials/custom-commit-notifications.md]: Perhaps `dispatch` or `send` rather than trigger? Trigger feels more appropriate for triggering events.

File: docs/tutorials/custom-commit-notifications.md [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+{!../../.atomist/handlers/event/NewCommitPushedToDMAndGeneral.ts!}
+```
+
+Event handlers can trigger multiple messages from a single event as shown 
+on the highlithed lines in this example.
+
+## Fetch and Display the Modified Files 
+
+Sending messages on events is critical for information flow but what if you 
+could perform actions on those events! On a commit, it could be handy to see
```
Comment (kipz) [inline docs/tutorials/custom-commit-notifications.md]: ! should be ?

File: docs/tutorials/custom-commit-notifications.md [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+
+## Fetch and Display the Modified Files 
+
+Sending messages on events is critical for information flow but what if you 
+could perform actions on those events! On a commit, it could be handy to see 
+which files were modified without having to open a browser or pulling the
+changes locally.
+
+Atomist allows your event handler to perform operations such as calling an 
+HTTP endpoint through instructions you add to the returned plan. Let's see
```
Comment (kipz) [inline docs/tutorials/custom-commit-notifications.md]: capitalise Instructions and Plan?

File: docs/tutorials/custom-commit-notifications.md [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+how this could look like with the following event handler.
+
+!!! note
+    This example relies on the [GitHub REST API][ghrestapi] and expects a public
```
Comment (kipz) [inline docs/tutorials/custom-commit-notifications.md]: s/a public access/anonymous access

File: docs/tutorials/custom-commit-notifications.md [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+    access to that repository for the call to succeed.
+
+[ghrestapi]: https://developer.github.com/v3/repos/commits/
+
+```typescript linenums="1"
+{!../../.atomist/handlers/event/ListCommitFilesChanged.ts!}
+```
+
+This handler is quite longer than in previous section indeed. The reason is that
```
Comment (kipz) [inline docs/tutorials/custom-commit-notifications.md]: s/quite/a bit/ or much etc.

File: docs/tutorials/custom-commit-notifications.md [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+it does much more as well and hopefully glancing at the code should make sense
```
Comment (kipz) [inline docs/tutorials/custom-commit-notifications.md]: `much more, and hopefully glancing at the code, should` - remove as well and add commas.

File: docs/tutorials/custom-commit-notifications.md [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+already.
+
+Indeed, on line 32, you instruct Atomist to perform a HTTP GET request against
```
Comment (kipz) [inline docs/tutorials/custom-commit-notifications.md]: `Indeed` is normally in response to something else - an agreement.
Maybe just `On line 32...`

File: docs/tutorials/custom-commit-notifications.md [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+the provided URL to retrieve the commit information, such as the number of
+files modified by that commit.
+
+!!! warning
+    Atomist has a [whitelist][] of services that can be accessed that way,
+    please contact us if you need to call a service not in that list.
+
+[whitelist]: https://github.com/atomist/rug-function-http/blob/master/src/com/atomist/rug/functions/rug_function_http/whitelist.clj
+
+The call is performed as soon as possible by Atomist which, in return calls the
```
Comment (kipz) [inline docs/tutorials/custom-commit-notifications.md]: remove comma, or add more commas `...Atomist, which, in return, calls...`

File: docs/tutorials/teach-atomist-new-skills.md [hunk 19]
```diff
@@ -0,0 +1,103 @@
+Software development does not occur solely on the developer's machine any
+longer. It has become a collaborative task opened to all members of the team
```
Comment (kipz) [inline docs/tutorials/teach-atomist-new-skills.md]: s/opened/open?

File: docs/tutorials/teach-atomist-new-skills.md [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+and interactions are constant.
```
Comment (kipz) [inline docs/tutorials/teach-atomist-new-skills.md]: continuous or on-going?

File: docs/tutorials/teach-atomist-new-skills.md [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+
+The Atomist Bot you have enrolled to your team strives to be tailored to what
+makes sense to your team. To do so, you can teach the bot new skills 
+automating common tasks your team routinely performs. Those skills are
```
Comment (kipz) [inline docs/tutorials/teach-atomist-new-skills.md]: s/automating/to automate

File: docs/tutorials/teach-atomist-new-skills.md [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+longer. It has become a collaborative task opened to all members of the team
+and interactions are constant.
+
+The Atomist Bot you have enrolled to your team strives to be tailored to what
+makes sense to your team. To do so, you can teach the bot new skills 
+automating common tasks your team routinely performs. Those skills are 
+implemented inside Rug command handlers and this tutorial will show you their 
+powerful, and indeed fun, capabilities.
+
+!!! tip "All you need is Rug"
```
Comment (kipz) [inline docs/tutorials/teach-atomist-new-skills.md]: This is duplicated from above - consider using an import?

File: docs/tutorials/teach-atomist-new-skills.md [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+    new [Rug command handlers][rugcmd]. You can easily add them using the 
+    [AddTypeScriptCommandHandler Rug editor][rugeditor] provided by Atomist.
+
+[ugpj]: /user-guide/rug/projects.md
+[rugcmd]: /user-guide/rug/commands.md
+[rugeditor]: https://github.com/atomist/rug-editors#addtypescriptcommandhandler
+
+## Search StackOverflow using the Bot
+
+StackOverflow is one of the biggest venue for developers to collaborate
```
Comment (kipz) [inline docs/tutorials/teach-atomist-new-skills.md]: s/biggest/most popular?

File: docs/tutorials/teach-atomist-new-skills.md [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+together. It is a rather common place to go to and using Atomist you can let
```
Comment (kipz) [inline docs/tutorials/teach-atomist-new-skills.md]: Remove `together` - kind of implied. Perhaps `collaborate when creating software`

Comment (kipz) [inline docs/tutorials/teach-atomist-new-skills.md]: s/to go/to hang out?

File: docs/tutorials/teach-atomist-new-skills.md [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+your team enjoy the service from the team's channel, interacting with each
```
Comment (kipz) [inline docs/tutorials/teach-atomist-new-skills.md]: `, whilst also interacting with each other.`? It reads a bit oddly otherwise.

File: docs/tutorials/teach-atomist-new-skills.md [hunk 26]
```diff
@@ -0,0 +1,103 @@
...
+other.
+
+Let's see a command handler you could use to query StackOverflow:
+
+```typescript linenums="1"
+{!../../.atomist/handlers/command/SearchStackOverflow.ts!}
+```
+
+Once published to your team, you would be able to make the most of this new
```
Comment (kipz) [inline docs/tutorials/teach-atomist-new-skills.md]: At this point, does `publish to your team` make sense?

File: docs/tutorials/teach-atomist-new-skills.md [hunk 27]
```diff
@@ -0,0 +1,103 @@
...
+skill as follows:
+
+```
+@atomist search SO
+```
+
+Upon which the Atomist Bot would prompt your to input your query before going
+its merry way querying StackOverflow. In return, it would disply your the first
```
Comment (kipz) [inline docs/tutorials/teach-atomist-new-skills.md]: `its merry way and querying`

File: docs/tutorials/teach-atomist-new-skills.md [hunk 28]
```diff
@@ -0,0 +1,103 @@
...
+returned items in a nicely formatted Slack message.
+
+Hpw does this work though? You define your Rug command handler and associate
```
Comment (kipz) [inline docs/tutorials/teach-atomist-new-skills.md]: `How`

File: docs/tutorials/teach-atomist-new-skills.md [hunk 29]
```diff
@@ -0,0 +1,103 @@
...
+
+This handler takes a single parameter called `query` holding the user-supplied
+input. Since this does not have a default value, the Atomist Bot will always
+ask for it.
+
+Then, line 23, we define the plan which Atomist will execute as soon as possible
+once the handler returns it. This plan instructs Atomist to call the
+StackOverflow REST API and use the `SendStackOverflowResults` response handler
+on success. That handler is defined on line 43 and simply renders the result
+content from the StackOverflow API call so that it renders nicely in the
```
Comment (kipz) [inline docs/tutorials/teach-atomist-new-skills.md]: Two uses of `renders`...maybe replace first with `process` or replace second with `looks nice`

File: docs/tutorials/teach-atomist-new-skills.md [hunk 30]
```diff
@@ -0,0 +1,103 @@
...
+channel you made the query from.
+
+!!! tip
+    Please refer to the [Rug command handler user-guide][rugcmd] to dive into
+    the machinery of those command handlers.
+
+## See Recent Active Issues
+
+In the morning, or when you start your day, you are likely interested in
+knowning what happend while you were away. Why not adding a new skill to your
```
Comment (kipz) [inline docs/tutorials/teach-atomist-new-skills.md]: `Why not add a new skill`

File: docs/tutorials/teach-atomist-new-skills.md [hunk 31]
```diff
@@ -0,0 +1,103 @@
...
+team so each member can ask for the last two or three active issues via the
+Atomist bot!
+
+Here is such a command handler for this new skill:
```
Comment (kipz) [inline docs/tutorials/teach-atomist-new-skills.md]: s/for/that implements?

File: docs/tutorials/teach-atomist-new-skills.md [hunk 32]
```diff
@@ -0,0 +1,103 @@
...
+{!../../.atomist/handlers/command/ListActiveIssues.ts!}
+```
+
+The Atomist Bot will run that command handler on the following intent:
+
+```
+@atomist active issues
+```
+
+The Bot will happily query the lits of issues but, as your project may
```
Comment (kipz) [inline docs/tutorials/teach-atomist-new-skills.md]: s/lits/list

File: docs/tutorials/teach-atomist-new-skills.md [hunk 33]
```diff
@@ -0,0 +1,103 @@
...
+be private or to work around the low limits of un-authenticated API call to
```
Comment (kipz) [inline docs/tutorials/teach-atomist-new-skills.md]: s/call/calls

File: docs/tutorials/teach-atomist-new-skills.md [hunk 34]
```diff
@@ -0,0 +1,103 @@
...
+the GitHub API, this command handler uses one of your GitHub token, line 11,
```
Comment (kipz) [inline docs/tutorials/teach-atomist-new-skills.md]: s/token/tokens

File: docs/tutorials/teach-atomist-new-skills.md [hunk 35]
```diff
@@ -0,0 +1,103 @@
...
+stored as a secret value by Atomist (line 36). This secret does not transit
+on the wire and is only available at runtime when the command handler is
+executed in the Atomist backend. You only need to declare it, line 11, so
```
Comment (kipz) [inline docs/tutorials/teach-atomist-new-skills.md]: s/in/on

File: docs/tutorials/teach-atomist-new-skills.md [hunk 36]
```diff
@@ -0,0 +1,103 @@
...
+    properly. Here, Atomist expects a token that has the `repo` scope
+    permission.
+
+[permissions]: /user-guide/permissions/github.md
+
+The command handler builds a plan instructing Atomist to query the GitHub API
+on your behalf. The plan also indicates how to process the response from that
+call (line 41).
+
+Indeed, the rendering of the returned issues is taken care by the
```
Comment (kipz) [inline docs/tutorials/teach-atomist-new-skills.md]: Remove `Indeed`.

File: .atomist/handlers/event/NewCommitPushedToDMAndGeneral.ts (hunk 1 above)
Comment (kipz) [inline .atomist/handlers/event/NewCommitPushedToDMAndGeneral.ts]: That's odd, because under the hood, we send it on a single field, so the bot is making an assumption about the way slack works, which may not hold in the future if we use other chat systems.

IMO - we should consistent here. Either we _always_ omit it in the constructor, and have the TS layer add it, or we should explicitly require it.

File: docs/tutorials/teach-atomist-new-skills.md (hunk 22 above)
Comment (Lawouach) [inline docs/tutorials/teach-atomist-new-skills.md]: actually, they don't link to the same editor ;)

File: .atomist/tests/handlers/event/NewCommitPushedToDMSteps.ts (hunk 4 above)
Comment (Lawouach) [inline .atomist/tests/handlers/event/NewCommitPushedToDMSteps.ts]: The repo is atomist/rug here.

File: .atomist/tests/handlers/command/ListActiveIssuesSteps.ts (hunk 3 above)
Comment (Lawouach) [inline .atomist/tests/handlers/command/ListActiveIssuesSteps.ts]: I'm not sure what you mean or what I should write differently.

File: .atomist/handlers/command/ListActiveIssues.ts [hunk 37]
```diff
@@ -0,0 +1,59 @@
...
+            instruction: {
+                kind: "execute",
+                name: "http",
+                parameters: {
+                    method: "get",
+                    url: api,
+                    config: {
+                        headers: {
+                            "Content-Type": "application/json",
+                            "Authorization": `token #{github://user_token?scopes=repo}`
```
Comment (ddgenome) [inline .atomist/handlers/command/ListActiveIssues.ts]: Is authorization required for public repos?

File: .atomist/tests/handlers/event/NewCommitPushedToDMSteps.ts (hunk 4 above)
Comment (ddgenome) [inline .atomist/tests/handlers/event/NewCommitPushedToDMSteps.ts]: I think that was a joke.

File: docs/tutorials/custom-commit-notifications.md [hunk 38]
```diff
@@ -0,0 +1,124 @@
+Things happen all the time in your development workflow. Commits are pushed,
+builds are started, they fail or succeed. Team members, or even users, open and
+comment issues everyday.
+
+All that flow of information can become overwhelming. With Atomist, you can
+attach behavior to these events and automate all sorts of near real-time
```
Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: I would omit "near real-time".

File: docs/tutorials/custom-commit-notifications.md [hunk 39]
```diff
@@ -0,0 +1,124 @@
...
+responses.
```
Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: "responses and actions."

File: docs/tutorials/custom-commit-notifications.md [hunk 40]
```diff
@@ -0,0 +1,124 @@
...
+
+These tutorials show you how to rely on Atomist to automate those scenarios.
```
Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: "how to use Atomist to automate"

File: docs/tutorials/custom-commit-notifications.md [hunk 41]
```diff
@@ -0,0 +1,124 @@
...
+
+!!! tip "All you need is Rug"
```
Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: Love it!

File: docs/tutorials/custom-commit-notifications.md [hunk 42]
```diff
@@ -0,0 +1,124 @@
...
+    These tutorials aim at showing you the code that you could write to
+    automate event reactions. They assume you have a [Rug project][ugpj]. You
+    can easily add [Rug event handler][rugev] to that project via the public 
+    [AddTypeScriptEventHandler Rug editor][rugeditor] provided by Atomist.
```
Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: "These tutorials focus on the code that you would write to
automate event reactions. They assume you have a [Rug project][ugpj] to add these files to. You
can easily add [Rug event handlers][rugev] to your project via the public 
[AddTypeScriptEventHandler Rug editor][rugeditor] provided by Atomist."

File: docs/tutorials/custom-commit-notifications.md [hunk 43]
```diff
@@ -0,0 +1,124 @@
...
+
+[rugev]: /user-guide/rug/handlers.md
+[ugpj]: /user-guide/rug/projects.md
```
Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: Why do you omit the leader `r`?

File: .atomist/tests/handlers/command/ListActiveIssuesSteps.ts (hunk 3 above)
Comment (ddgenome) [inline .atomist/tests/handlers/command/ListActiveIssuesSteps.ts]: The tests are not actually shown or discussed in the tutorial text, so perhaps this discussion can be had offline.

File: docs/tutorials/custom-commit-notifications.md [hunk 44]
```diff
@@ -0,0 +1,124 @@
...
+[ugpub]: /user-guide/rug/lifecycle.md#publishing
+
+Below is such an event handler:
+
+```typescript linenums="1"
+{!../../.atomist/handlers/event/NewCommitPushedToGeneral.ts!}
+```
+
+An event handler tells Atomist what it is interested in via its
+[path expression][ugpxe] on line 10. Notice in particular how the handler
```
Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: Lines 11-13 when I build the site. I would reword this a bit to "via its path expression, which we see on lines 11--13".

File: docs/tutorials/custom-commit-notifications.md (hunk 8 above)
Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: line 12

Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: Note there is a difference between the `name` and `login` properties on `GitHubId`. The former is not unique and is free text. We may want to use `login` for this example.

File: docs/tutorials/custom-commit-notifications.md [hunk 45]
```diff
@@ -0,0 +1,124 @@
...
+filters only commit made by a given author on line 11 (here, Alice's commits). 
+Finally, line 12 asks Atomist to also feed the handler with the commits
```
Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: line 13

File: docs/tutorials/custom-commit-notifications.md [hunk 46]
```diff
@@ -0,0 +1,124 @@
...
+{!../../.atomist/handlers/event/NewCommitPushedToDM.ts!}
+```
+
+As you can see, the only change is on line 22 where you provide the name of 
+the user to send the message to, here `@Bob`.
+
+But wait, you can have it both ways and send a direct message and a 
+channel message:
+
+```typescript linenums="1" hl_lines="21 26 31 33"
```
Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: I would _not_ highlight lines 21 and 33.

File: docs/tutorials/custom-commit-notifications.md [hunk 47]
```diff
@@ -0,0 +1,124 @@
...
+{!../../.atomist/handlers/event/NewCommitPushedToDMAndGeneral.ts!}
+```
+
+Event handlers can dispatch multiple messages from a single event as shown 
+on the highlithed lines in this example.
+
+## Fetch and Display the Modified Files
```
Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: Perhaps this should be moved to a new page/tutorial?

Thoughts @WillGardella ?

File: mkdocs.yml [hunk 48]
```diff
@@ -70,8 +70,8 @@ pages:
     - What next: getting-started/what-next.md
 - Tutorials:
     - Introduction: tutorials/index.md
-    - Build a Handler: tutorials/handler.md
-    - Build a Generator: tutorials/generator.md
+    - Custom Commit Notifications: tutorials/custom-commit-notifications.md
+    - Teach your Team new Skills: tutorials/teach-atomist-new-skills.md
```
Comment (ddgenome) [inline mkdocs.yml]: I would change the page title to "Teach Atomist New Skills".

File: docs/tutorials/custom-commit-notifications.md [hunk 49]
```diff
@@ -0,0 +1,124 @@
...
+{!../../.atomist/handlers/event/NewCommitPushedToDMAndGeneral.ts!}
+```
+
+Event handlers can dispatch multiple messages from a single event as shown 
+on the highlithed lines in this example.
+
+## Fetch and Display the Modified Files 
+
+Sending messages on events is critical for information flow but what if you 
+could perform actions on those events! On a commit, it should be handy to see
```
Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: You are still just sending a message with this handler, albeit a message with more information, so the first sentence here is a bit misleading. The first sentence could be "Sending simple commit notifications is useful, but can we provide more context so people can get all the information they need without leaving Slack and go to GitHub.com?"

File: docs/tutorials/custom-commit-notifications.md [hunk 50]
```diff
@@ -0,0 +1,124 @@
...
+## Fetch and Display the Modified Files 
+
+Sending messages on events is critical for information flow but what if you 
+could perform actions on those events! On a commit, it should be handy to see 
+which files were modified without having to open a browser or pulling the
+changes locally.
+
+Atomist allows your event handler to perform operations such as calling an 
+HTTP endpoint through Instructions you add to the returned Plan. Let's see
+how this could look like with the following event handler.
```
Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: I'd combine these first two paragraphs.

File: docs/tutorials/custom-commit-notifications.md [hunk 51]
```diff
@@ -0,0 +1,124 @@
...
+
+[ghrestapi]: https://developer.github.com/v3/repos/commits/
+
+```typescript linenums="1"
+{!../../.atomist/handlers/event/ListCommitFilesChanged.ts!}
+```
+
+This handler is a bit longer than in previous section indeed. The reason is that
+it does much more, and hopefully glancing at the code should make sense
+already.
```
Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: This paragraph does not really add anything.

Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: I think it would be better to discuss how this handler uses callbacks, i.e., ResponseHandlers, so it can use the information from the asynchronously called HTTP action.

File: docs/tutorials/custom-commit-notifications.md [hunk 52]
```diff
@@ -0,0 +1,124 @@
...
+
+On line 32, you instruct Atomist to perform a HTTP GET request against
+the provided URL to retrieve the commit information, such as the number of
```
Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: "such as" is not correct here, "specifically" would be accurate.

File: docs/tutorials/custom-commit-notifications.md [hunk 53]
```diff
@@ -0,0 +1,124 @@
...
+files modified by that commit.
+
+!!! warning
+    Atomist has a [whitelist][] of services that can be accessed that way,
+    please contact us if you need to call a service not in that list.
+
+[whitelist]: https://github.com/atomist/rug-function-http/blob/master/src/com/atomist/rug/functions/rug_function_http/whitelist.clj
```
Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: Probably better to link to the README, which also includes the white list, as the location of this file may change.

File: docs/tutorials/custom-commit-notifications.md [hunk 54]
```diff
@@ -0,0 +1,124 @@
...
+
+The call is performed as soon as possible by Atomist, which, in return, calls
```
Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: `s/as soon as possible/asynchronously/`

Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: `s/which, in turn,/which, after completing the HTTP call,/`

File: docs/tutorials/custom-commit-notifications.md [hunk 55]
```diff
@@ -0,0 +1,124 @@
...
+
+```typescript linenums="1"
+{!../../.atomist/handlers/event/ListCommitFilesChanged.ts!}
+```
+
+This handler is a bit longer than in previous section indeed. The reason is that
+it does much more, and hopefully glancing at the code should make sense
+already.
+
+On line 32, you instruct Atomist to perform a HTTP GET request against
```
Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: "On lines 29--54 you instruct"

File: docs/tutorials/custom-commit-notifications.md [hunk 56]
```diff
@@ -0,0 +1,124 @@
...
+[whitelist]: https://github.com/atomist/rug-function-http/blob/master/src/com/atomist/rug/functions/rug_function_http/whitelist.clj
+
+The call is performed as soon as possible by Atomist, which, in return, calls
+the appropriate success or error handler based on the response from the HTTP
+call.
+
+Here, you indicate that errors should be simply returned as a generic error 
+message using the `GenericErrorHandler` provided by Atomist. However, succesful
+calls will be handled by the `ReceivedCommitDetails` response handler defined
+line 61.
```
Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: "lines 61--98."

File: docs/tutorials/custom-commit-notifications.md [hunk 57]
```diff
@@ -0,0 +1,124 @@
...
+
+!!! tip
+    This response handler should live in the same TypeScript file as its event
+    handler.
+
+
```
Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: extra empty line

File: docs/tutorials/custom-commit-notifications.md [hunk 58]
```diff
@@ -0,0 +1,124 @@
...
+Your response handler prepares a [Slack-ready message][slack] with the changes
```
Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: `s/Slack-ready/chat/`.

File: docs/tutorials/custom-commit-notifications.md [hunk 59]
```diff
@@ -0,0 +1,124 @@
...
+from that commit. Notice how the response from the HTTP call is automatically
+parsed, line 63, for you and available via `response.body` as an object.
```
Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]:
```
parsed using the `#!typescript @ParseJson` annotation on line 63 and
make available as an object via `#!typescript response.body`.
```

File: docs/tutorials/custom-commit-notifications.md (hunk 58 above)
Comment (ddgenome) [inline docs/tutorials/custom-commit-notifications.md]: remove slack link

**Comment by Lawouach** [comment]
I'm afraid I won't be able able to fix those changes so if they are important, please, by all means, just make those changes directly :)

File: mkdocs.yml (hunk 48 above)
Comment (Lawouach) [inline mkdocs.yml]: I wanted to stress you improve your team, not Atomist. I thought our message was meant to say "Atomust majes you better", not "Make the Bot better".

**Comment by ddgenome** [comment]
I'm working through this. Rather than commenting again, I am just going to fix, commit, and publish.
